File: src/apis/sources/youtube.mjs
// Nepali broadcast video for the flood desk.
//
// Television is where a lot of this coverage actually lives — a correspondent
// standing where a bridge used to be says more than the wire copy about it —
// and the Nepali channels put it on YouTube within the hour.
//
// How Atlas reaches it, and why:
//
//   Embed, never download. Videos are shown in YouTube's own iframe player, so
//   the outlet keeps its view count and its advertising. Nothing is copied to
//   MinIO; that bucket is for photographs the public sent us and nothing else.
//
//   The Data API when a key exists. YOUTUBE_API_KEY gives proper search, which
//   is the only way to find relevant video across channels Atlas has not been
//   told about.
//
//   Channel pages and oEmbed when it does not. YouTube's old RSS endpoint
//   (feeds/videos.xml) now answers 404, so recent video ids are read off the
//   channel page and each one's title, author and thumbnail come from the
//   documented oEmbed endpoint rather than from scraped markup.

import { safeFetch } from '../utils/fetch.mjs';

const BROWSER_UA =
  'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0 Safari/537.36';

/**
 * Nepali news channels, verified by channel id.
 *
 * Handles change and get squatted; ids do not. To add one, open the channel and
 * take the UC… id out of its /channel/ URL.
 */
export const DEFAULT_CHANNELS = [
  { id: 'UC3yDoaqQzOd1bNP74ZrGPTA', name: 'Kantipur TV HD' },
  { id: 'UCo4cuctdb-1YdZNgWEVZGwA', name: 'Onlinekhabar' },
  { id: 'UCjG2HX7jfwqIjzTlaF1CPGA', name: 'News24 Nepal' },
  { id: 'UCg04d_B2CR0Di1rs7mSu6JQ', name: 'Nepal News' },
];

// A video is only shown if its title says it is about this event. Broadcasters
// post everything to the same channel, and a flood desk carrying last night's
// football highlights would be worse than carrying no video at all.
const RELEVANCE_TERMS = [
  'बाढी', 'पहिरो', 'रसुवा', 'भोटेकोशी', 'त्रिशूली', 'उद्धार', 'राहत', 'विपद्',
  'डुबान', 'बगायो', 'सखाप', 'घाइते', 'बेपत्ता', 'तटबन्ध', 'हिमताल', 'लेन्डे',
  'नुवाकोट', 'धादिङ', 'तिमुरे', 'स्याफ्रुबेसी', 'धुन्चे', 'बेत्रावती',
  'flood', 'landslide', 'rasuwa', 'bhotekoshi', 'bhote koshi', 'trishuli',
  'rescue', 'relief', 'disaster', 'washed away', 'timure', 'syafrubesi',
  'dhunche', 'nuwakot', 'dhading', 'glof', 'inundat',
];

const TTL_MS = 10 * 60 * 1000;
const cache = new Map();

async function cached(key, loader) {
  const hit = cache.get(key);
  if (hit && Date.now() - hit.at < TTL_MS) return hit.value;
  const value = await loader();
  cache.set(key, { value, at: Date.now() });
  return value;
}

function isRelevant(text) {
  const lower = String(text || '').toLowerCase();
  return RELEVANCE_TERMS.some(term => lower.includes(term.toLowerCase()));
}

/**
 * Recent video ids from a channel's uploads page.
 *
 * The ids appear in the bootstrap JSON the page ships with. Only the eleven
 * character ids are taken — no titles, no descriptions, nothing that oEmbed
 * will give authoritatively a moment later.
 */
async function getChannelVideoIds(channelId, limit = 25) {
  const res = await fetch(`https://www.youtube.com/channel/${channelId}/videos`, {
    signal: AbortSignal.timeout(20_000),
    headers: { 'User-Agent': BROWSER_UA, 'Accept-Language': 'ne,en;q=0.8' },
  });
  if (!res.ok) throw new Error(`YouTube channel HTTP ${res.status}`);
  const html = await res.text();

  const ids = [];
  const seen = new Set();
  for (const match of html.matchAll(/"videoId":"([A-Za-z0-9_-]{11})"/g)) {
    const id = match[1];
    if (seen.has(id)) continue;
    seen.add(id);
    ids.push(id);
    if (ids.length >= limit) break;
  }
  return ids;
}

/** Title, author and thumbnail for one video, from YouTube's oEmbed endpoint. */
async function getVideoMeta(videoId) {
  return cached(`meta:${videoId}`, async () => {
    const target = encodeURIComponent(`https://www.youtube.com/watch?v=${videoId}`);
    const data = await safeFetch(`https://www.youtube.com/oembed?url=${target}&format=json`, {
      timeout: 12_000,
      retries: 1,
    });
    if (!data || data.error || !data.title) return null;
    return {
      id: videoId,
      title: data.title,
      channel: data.author_name || null,
      channelUrl: data.author_url || null,
      thumbnail: data.thumbnail_url || `https://i.ytimg.com/vi/${videoId}/hqdefault.jpg`,
      url: `https://www.youtube.com/watch?v=${videoId}`,
      embedUrl: `https://www.youtube-nocookie.com/embed/${videoId}`,
      publishedAt: null,
    };
  });
}

/** Search, when a Data API key is configured. Returns [] when it is not. */
async function searchViaApi(query, limit) {
  const key = process.env.YOUTUBE_API_KEY;
  if (!key) return [];
  const params = new URLSearchParams({
    part: 'snippet',
    q: query,
    type: 'video',
    order: 'date',
    maxResults: String(Math.min(limit, 50)),
    relevanceLanguage: 'ne',
    regionCode: 'NP',
    key,
  });
  const data = await safeFetch(`https://www.googleapis.com/youtube/v3/search?${params}`, { timeout: 15_000 });
  if (!data || data.error) {
    console.warn('[YouTube] Data API search failed:', data?.error?.message || data?.error);
    return [];
  }
  return (data.items || [])
    .filter(item => item.id?.videoId)
    .map(item => ({
      id: item.id.videoId,
      title: item.snippet?.title || '',
      channel: item.snippet?.channelTitle || null,
      channelUrl: item.snippet?.channelId ? `https://www.youtube.com/channel/${item.snippet.channelId}` : null,
      thumbnail: item.snippet?.thumbnails?.high?.url || `https://i.ytimg.com/vi/${item.id.videoId}/hqdefault.jpg`,
      url: `https://www.youtube.com/watch?v=${item.id.videoId}`,
      embedUrl: `https://www.youtube-nocookie.com/embed/${item.id.videoId}`,
      publishedAt: item.snippet?.publishedAt || null,
    }));
}

async function getLiveVideoId(channelId) {
  try {
    const res = await fetch(`https://www.youtube.com/channel/${channelId}/live`, {
      signal: AbortSignal.timeout(10_000),
      headers: { 'User-Agent': BROWSER_UA }
    });
    if (!res.ok) return null;
    const html = await res.text();
    if (!html.includes('"isLive":true')) return null;
    
    const canonicalMatch = html.match(/<link rel="canonical" href="https:\/\/www\.youtube\.com\/watch\?v=([A-Za-z0-9_-]{11})">/);
    if (canonicalMatch) return canonicalMatch[1];
    
    const match = html.match(/"videoId":"([A-Za-z0-9_-]{11})"/);
    return match ? match[1] : null;
  } catch (e) {
    return null;
  }
}

/**
 * Flood coverage from the Nepali broadcasters.
 *
 * @param {object} [opts]
 * @param {Array<{id: string, name: string}>} [opts.channels]
 * @param {number} [opts.limit]
 * @param {string} [opts.query] search string, used only when a Data API key is set
 */
export async function getFloodVideos({ channels = DEFAULT_CHANNELS, limit = 24, query = 'रसुवा बाढी भोटेकोशी' } = {}) {
  const fetchedAt = new Date().toISOString();
  const errors = [];

  const liveChannels = [
    { id: 'UC3yDoaqQzOd1bNP74ZrGPTA', name: 'Kantipur TV HD Live', channelName: 'Kantipur TV' },
    { id: 'UCjG2HX7jfwqIjzTlaF1CPGA', name: 'News24 Nepal Live', channelName: 'News24' },
    { id: 'UCNye-w-v_C-fjh7W023RypQ', name: 'Al Jazeera English Live', channelName: 'Al Jazeera' },
    { id: 'UCknLrEdhRCp1gqcb7cyYA_Q', name: 'DW News Live', channelName: 'DW News' },
    { id: 'UCwqusr8YDwM-3mEYTDeJHzw', name: 'Republic World Live', channelName: 'Republic World' },
    { id: 'UC83iGbaOhZR8AWNYYNzSNjg', name: 'India Global Review Live', channelName: 'India Global Review' }
  ];

  const [apiResults, channelIdLists, liveStreamsResults] = await Promise.all([
    searchViaApi(query, limit).catch(err => {
      errors.push(`search: ${err.message}`);
      return [];
    }),
    Promise.allSettled(channels.map(c => getChannelVideoIds(c.id))),
    Promise.allSettled(liveChannels.map(async c => {
      const vidId = await getLiveVideoId(c.id);
      if (!vidId) return null;
      return {
        id: vidId,
        title: c.name,
        channel: c.channelName,
        channelUrl: `https://www.youtube.com/channel/${c.id}`,
        thumbnail: `https://i.ytimg.com/vi/${vidId}/hqdefault.jpg`,
        url: `https://www.youtube.com/watch?v=${vidId}`,
        embedUrl: `https://www.youtube.com/embed/${vidId}`,
        publishedAt: null
      };
    }))
  ]);

  const live = [];
  for (const r of liveStreamsResults) {
    if (r.status === 'fulfilled' && r.value) live.push(r.value);
  }

  const ids = [];
  channelIdLists.forEach((result, i) => {
    if (result.status === 'fulfilled') ids.push(...result.value);
    else errors.push(`${channels[i].name}: ${result.reason?.message || result.reason}`);
  });

  // oEmbed is one request per video, so cap the fan-out and lean on the cache.
  const uniqueIds = [...new Set(ids)].slice(0, 60);
  const metas = [];
  const CONCURRENCY = 8;
  for (let i = 0; i < uniqueIds.length; i += CONCURRENCY) {
    const slice = uniqueIds.slice(i, i + CONCURRENCY);
    const settled = await Promise.allSettled(slice.map(getVideoMeta));
    for (const r of settled) if (r.status === 'fulfilled' && r.value) metas.push(r.value);
  }

  // Search results first — they were matched on the query, not merely on being
  // recent — then channel uploads whose titles are about this event.
  const merged = [...apiResults, ...metas];
  const seen = new Set();
  const videos = [];
  for (const video of merged) {
    if (seen.has(video.id)) continue;
    if (!isRelevant(video.title)) continue;
    seen.add(video.id);
    videos.push(video);
    if (videos.length >= limit) break;
  }

  return {
    videos,
    live,
    searchEnabled: Boolean(process.env.YOUTUBE_API_KEY),
    error: errors.length ? errors.join('; ') : null,
    fetchedAt,
  };
}

File: backend/app/domains/news/sources/youtube.py
"""Nepali broadcast video for the flood desk.

Television is where a lot of this coverage actually lives — a correspondent
standing where a bridge used to be says more than the wire copy about it — and
the Nepali channels put it on YouTube within the hour.

How Atlas reaches it, and why:

  EMBED, NEVER DOWNLOAD. Videos are shown in YouTube's own iframe player, so the
  outlet keeps its view count and its advertising. Nothing is copied to MinIO;
  that bucket is for photographs the public sent us and nothing else.

  The Data API when a key exists. YOUTUBE_API_KEY gives proper search, which is
  the only way to find relevant video across channels Atlas has not been told
  about.

  Channel pages and oEmbed when it does not. YouTube's old RSS endpoint answers
  404 now, so recent video ids are read off the channel page and each one's
  title, author and thumbnail come from the documented oEmbed endpoint rather
  than from scraped markup.
"""

import asyncio
import re
import time
from typing import Any, NamedTuple
from urllib.parse import quote

from app.core.config import settings
from app.core.http import is_error, now_iso, safe_fetch
from app.core.logging import get_logger

log = get_logger(__name__)

BROWSER_UA = (
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/120.0 Safari/537.36"
)


class Channel(NamedTuple):
    id: str
    name: str


# Nepali news channels, verified by channel id. Handles change and get
# squatted; ids do not.
DEFAULT_CHANNELS = [
    Channel("UC3yDoaqQzOd1bNP74ZrGPTA", "Kantipur TV HD"),
    Channel("UCo4cuctdb-1YdZNgWEVZGwA", "Onlinekhabar"),
    Channel("UCjG2HX7jfwqIjzTlaF1CPGA", "News24 Nepal"),
    Channel("UCg04d_B2CR0Di1rs7mSu6JQ", "Nepal News"),
]

# Nepali broadcasters only.
#
# The list used to carry Al Jazeera, DW, Republic World and India Global
# Review. A rolling international channel is almost never on this flood when a
# reader opens it — it is on whatever the world is covering that hour — so it
# took space from the broadcasters actually reporting from the corridor and
# lent a foreign anchor the authority of appearing on a Nepal flood desk.
LIVE_CHANNELS = [
    ("UC3yDoaqQzOd1bNP74ZrGPTA", "Kantipur TV HD Live", "Kantipur TV"),
    ("UCjG2HX7jfwqIjzTlaF1CPGA", "News24 Nepal Live", "News24"),
    ("UCo4cuctdb-1YdZNgWEVZGwA", "Onlinekhabar Live", "Onlinekhabar"),
    ("UCg04d_B2CR0Di1rs7mSu6JQ", "Nepal News Live", "Nepal News"),
]

# A video is only shown if its title says it is about this event. Broadcasters
# post everything to the same channel, and a flood desk carrying last night's
# football highlights would be worse than carrying no video at all.
RELEVANCE_TERMS = [
    "बाढी", "पहिरो", "रसुवा", "भोटेकोशी", "त्रिशूली", "उद्धार", "राहत", "विपद्",
    "डुबान", "बगायो", "सखाप", "घाइते", "बेपत्ता", "तटबन्ध", "हिमताल", "लेन्डे",
    "नुवाकोट", "धादिङ", "तिमुरे", "स्याफ्रुबेसी", "धुन्चे", "बेत्रावती",
    "flood", "landslide", "rasuwa", "bhotekoshi", "bhote koshi", "trishuli",
    "rescue", "relief", "disaster", "washed away", "timure", "syafrubesi",
    "dhunche", "nuwakot", "dhading", "glof", "inundat",
]

TTL_S = 10 * 60
_cache: dict[str, tuple[float, Any]] = {}

_VIDEO_ID = re.compile(r'"videoId":"([A-Za-z0-9_-]{11})"')
_CANONICAL = re.compile(
    r'<link rel="canonical" href="https://www\.youtube\.com/watch\?v=([A-Za-z0-9_-]{11})">'
)


async def _cached(key: str, loader: Any) -> Any:
    hit = _cache.get(key)
    if hit and (time.monotonic() - hit[0]) < TTL_S:
        return hit[1]
    value = await loader()
    _cache[key] = (time.monotonic(), value)
    return value


def is_relevant(text: str | None) -> bool:
    lower = str(text or "").lower()
    return any(term.lower() in lower for term in RELEVANCE_TERMS)


async def get_channel_video_ids(channel_id: str, limit: int = 25) -> list[str]:
    """Recent video ids from a channel's uploads page.

    Only the eleven-character ids are taken — no titles, no descriptions,
    nothing that oEmbed will give authoritatively a moment later.
    """
    html = await safe_fetch(
        f"https://www.youtube.com/channel/{channel_id}/videos",
        as_="text",
        timeout=20.0,
        retries=0,
        headers={"User-Agent": BROWSER_UA, "Accept-Language": "ne,en;q=0.8"},
    )
    if not isinstance(html, str):
        raise RuntimeError("YouTube channel page unavailable")

    ids: list[str] = []
    seen: set[str] = set()
    for match in _VIDEO_ID.finditer(html):
        video_id = match.group(1)
        if video_id in seen:
            continue
        seen.add(video_id)
        ids.append(video_id)
        if len(ids) >= limit:
            break
    return ids


async def get_video_meta(video_id: str) -> dict[str, Any] | None:
    """Title, author and thumbnail, from YouTube's documented oEmbed endpoint."""

    async def load() -> dict[str, Any] | None:
        target = quote(f"https://www.youtube.com/watch?v={video_id}", safe="")
        data = await safe_fetch(
            f"https://www.youtube.com/oembed?url={target}&format=json",
            timeout=12.0,
            retries=1,
        )
        if is_error(data) or not isinstance(data, dict) or not data.get("title"):
            return None
        return {
            "id": video_id,
            "title": data["title"],
            "channel": data.get("author_name"),
            "channelUrl": data.get("author_url"),
            "thumbnail": data.get("thumbnail_url")
            or f"https://i.ytimg.com/vi/{video_id}/hqdefault.jpg",
            "url": f"https://www.youtube.com/watch?v={video_id}",
            # nocookie: the reader has not asked YouTube for anything yet.
            "embedUrl": f"https://www.youtube-nocookie.com/embed/{video_id}",
            "publishedAt": None,
        }

    return await _cached(f"meta:{video_id}", load)


async def search_via_api(query: str, limit: int) -> list[dict[str, Any]]:
    """Search, when a Data API key is configured. Empty list when it is not."""
    key = settings.YOUTUBE_API_KEY
    if not key:
        return []

    data = await safe_fetch(
        "https://www.googleapis.com/youtube/v3/search",
        timeout=15.0,
        params={
            "part": "snippet",
            "q": query,
            "type": "video",
            "order": "date",
            "maxResults": min(limit, 50),
            "relevanceLanguage": "ne",
            "regionCode": "NP",
            "key": key,
        },
    )
    if is_error(data) or not isinstance(data, dict) or data.get("error"):
        log.warning("youtube_search_failed")
        return []

    out = []
    for item in data.get("items") or []:
        video_id = (item.get("id") or {}).get("videoId")
        if not video_id:
            continue
        snippet = item.get("snippet") or {}
        channel_id = snippet.get("channelId")
        thumbnails = snippet.get("thumbnails") or {}
        out.append(
            {
                "id": video_id,
                "title": snippet.get("title") or "",
                "channel": snippet.get("channelTitle"),
                "channelUrl": f"https://www.youtube.com/channel/{channel_id}"
                if channel_id
                else None,
                "thumbnail": (thumbnails.get("high") or {}).get("url")
                or f"https://i.ytimg.com/vi/{video_id}/hqdefault.jpg",
                "url": f"https://www.youtube.com/watch?v={video_id}",
                "embedUrl": f"https://www.youtube-nocookie.com/embed/{video_id}",
                "publishedAt": snippet.get("publishedAt"),
            }
        )
    return out


async def get_live_video_id(channel_id: str) -> str | None:
    html = await safe_fetch(
        f"https://www.youtube.com/channel/{channel_id}/live",
        as_="text",
        timeout=10.0,
        retries=0,
        headers={"User-Agent": BROWSER_UA},
    )
    if not isinstance(html, str) or '"isLive":true' not in html:
        return None
    canonical = _CANONICAL.search(html)
    if canonical:
        return canonical.group(1)
    match = _VIDEO_ID.search(html)
    return match.group(1) if match else None


async def get_flood_videos(
    channels: list[Channel] | None = None,
    limit: int = 24,
    query: str = "रसुवा बाढी भोटेकोशी",
) -> dict[str, Any]:
    """Flood coverage from the Nepali broadcasters."""
    channels = channels or DEFAULT_CHANNELS
    fetched_at = now_iso()
    errors: list[str] = []

    async def live_for(entry: tuple[str, str, str]) -> dict[str, Any] | None:
        channel_id, title, channel_name = entry
        video_id = await get_live_video_id(channel_id)
        if not video_id:
            return None
        return {
            "id": video_id,
            "title": title,
            "channel": channel_name,
            "channelUrl": f"https://www.youtube.com/channel/{channel_id}",
            "thumbnail": f"https://i.ytimg.com/vi/{video_id}/hqdefault.jpg",
            "url": f"https://www.youtube.com/watch?v={video_id}",
            "embedUrl": f"https://www.youtube.com/embed/{video_id}",
            "publishedAt": None,
        }

    api_results, id_lists, live_results = await asyncio.gather(
        search_via_api(query, limit),
        asyncio.gather(
            *(get_channel_video_ids(c.id) for c in channels), return_exceptions=True
        ),
        asyncio.gather(*(live_for(c) for c in LIVE_CHANNELS), return_exceptions=True),
    )

    live = [r for r in live_results if isinstance(r, dict)]

    ids: list[str] = []
    for channel, result in zip(channels, id_lists):
        if isinstance(result, list):
            ids.extend(result)
        else:
            errors.append(f"{channel.name}: {result}")

    # oEmbed is one request per video, so cap the fan-out and lean on the cache.
    unique_ids = list(dict.fromkeys(ids))[:60]
    metas: list[dict[str, Any]] = []
    concurrency = 8
    for i in range(0, len(unique_ids), concurrency):
        chunk = unique_ids[i : i + concurrency]
        settled = await asyncio.gather(
            *(get_video_meta(v) for v in chunk), return_exceptions=True
        )
        metas.extend(r for r in settled if isinstance(r, dict))

    # Search results first — they were matched on the query, not merely on
    # being recent — then channel uploads whose titles are about this event.
    seen: set[str] = set()
    videos: list[dict[str, Any]] = []
    for video in [*api_results, *metas]:
        if video["id"] in seen or not is_relevant(video["title"]):
            continue
        seen.add(video["id"])
        videos.append(video)
        if len(videos) >= limit:
            break

    return {
        "videos": videos,
        "live": live,
        "searchEnabled": bool(settings.YOUTUBE_API_KEY),
        "error": "; ".join(errors) if errors else None,
        "fetchedAt": fetched_at,
    }
